docs/configuration.md
# Brawl Server Configuration

This file refers to the `config.toml` used by the server. If you are looking for a repository configuration reference, please refer to the [**`repository configuration reference`**](./repository-configuration.md).

## Level

The `level` option is used to set the logging level. This can be one of the following values:

- `trace`
- `debug`
- `info`
- `warn`
- `error`

It can also be a compound value using the tracing [**`env-filter syntax`**](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#directives).

## Telemetry Bind

The `telemetry_bind` option is used to set the bind address for the telemetry server. This is useful if you want to run the server in a container and want to expose the telemetry endpoint.

By default this is not set, so the telemetry server is not started.

However if the telemetry address is provided a server will listen on the address and expose a `/metrics` endpoint as well as a `/health` endpoint.

The health endpoint runs a check to see if the database is healthy and can be reached.

The metrics endpoint exposes prometheus metrics used for monitoring the server.

## Database URL

The `db_url` option is used to set the database URL. This is the URL of the database that brawl will use to store its data.

This defaults to the `DATABASE_URL` environment variable, if it is not set brawl will not be able to run.

This should be set to a valid postgres database url; for example:

```
DATABASE_URL=postgres://brawl:brawl@localhost:5432/brawl
```

## Interval Seconds

The `interval_seconds` option is a refresh rate used by brawl to refresh PRs and check for timeouts.

This defaults to 30 seconds.

## GitHub

The `github` section is used to configure the GitHub app.

This section is required.

### Webhook Bind

The `webhook_bind` is a bind address to run the webhook server on.

This defaults to `0.0.0.0:3000`.

### App ID

The `app_id` is the ID of the GitHub app.


This is required.

### Private Key PEM

The `private_key_pem` is the private key of the GitHub app. This should not be encoded in base64, it should be the raw private key.

You can generate a private key by going to the GitHub App settings page under the `General` tab and scrolling down to the `Private key` section. For more information look at the [**`Setup Guide`**](./setup.md#5-generate-a-private-key).

This is required.

### Webhook Secret

The `webhook_secret` is the secret used to verify the webhook.

This can be configured on the GitHub App settings page under the `Webhook` section. For more information look at the [**`Setup Guide`**](./setup.md#2-provide-a-webhook-url).

This is required.


docs/repository-configuration.md
# Repository Configuration

This file refers to the `.github/brawl.toml` file used to configure brawl for a given repository. If you are looking for a server configuration reference, please refer to the [**`server configuration reference`**](./configuration.md).

The configuration here is for the [**`GitHubBrawlRepoConfig`**](../server/src/github/config.rs) struct.

## Configuration Reference

### Enabled

The `enabled` option is a boolean that determines if brawl is enabled for this repository.

This defaults to `true`.

### Labels

The `labels` section is used to specify the labels to attach to PRs on different states.

#### On Merge Queued

The `on_merge_queued` option is a list of labels to attach to PRs when they are in the merge queue.

This defaults to an empty list.

These labels will be added to a PR when a it has been added to the merge queue before the merge has started.

#### On Merge In Progress

The `on_merge_in_progress` option is a list of labels to attach to PRs when they are being merged.

This defaults to an empty list.

These labels will be added to a PR when a it has started merging.

#### On Merge Failure

The `on_merge_failure` option is a list of labels to attach to PRs when they fail to merge.

This defaults to an empty list.

These labels will be added to a PR when a it has failed to merge.

#### On Merge Success

The `on_merge_success` option is a list of labels to attach to PRs when they are merged.

This defaults to an empty list.

These labels will be added to a PR when a it has been merged.

#### On Try In Progress

The `on_try_in_progress` option is a list of labels to attach to PRs when they are being tried.

This defaults to an empty list.

These labels will be added to a PR when a it has started trying.

#### On Try Failure

The `on_try_failure` option is a list of labels to attach to PRs when they fail to try.

This defaults to an empty list.

These labels will be added to a PR when a it has failed to try.

#### Auto-try Enabled

The `auto_try_enabled` option is a list of labels to attach to PRs when auto-try is enabled.

This defaults to an empty list.

These labels will be added to a PR when auto-try is enabled.

### Branches

The `branches` option is a list of branches that brawl will listen to.

This defaults to an empty list.

These branches will be used to determine if a PR should be merged or tried.

### Try Branch Prefix

The `try_branch_prefix` option is the prefix for the branch name used when trying a PR.

This defaults to `automation/brawl/try/`.

When creating a temporary branch for trying a PR, brawl will use this prefix in combination with the PR number to create the branch name.

For example, if the PR number is `123` and the prefix is `automation/brawl/try/`, the branch name will be `automation/brawl/try/123`.

Brawl will strip any trailing `/` from the prefix and always combine it with the PR number. So `automation/brawl/try/` and `automation/brawl/try` are the same.

### Merge Branch Prefix

The `merge_branch_prefix` option is the prefix for the branch name used when merging a PR.

This defaults to `automation/brawl/merge/`.

When creating a temporary branch for merging a PR, brawl will use this prefix in combination with the target branch name to create the branch name.

For example, if the target branch is `main` and the prefix is `automation/brawl/merge/`, the branch name will be `automation/brawl/merge/main`.

Brawl will strip any trailing `/` from the prefix and always combine it with the target branch name. So `automation/brawl/merge/` and `automation/brawl/merge` are the same.

### Temp Branch Prefix

The `temp_branch_prefix` option is the prefix for the branch name used when creating a temporary branch for help in creating a merge commit. Due to the limitations of the GitHub API we need to create a temporary branch to create a merge commit. Technically we could get around this by using the git api directly but we avoid using `git` for now to keep the process simple.

This defaults to `automation/brawl/temp/`.

So when we are doing either a merge or a try, we will create a temporary branch in the form of `automation/brawl/temp/<uuid>` and then perform a push to push this branch to the base and then merge the head into the base to create a merge commit and then delete the branch.

This branch is never persisted and is deleted immediately after we get a commit sha.

### Merge Permissions

The `merge_permissions` option is a list of permissions required to merge a PR.

This defaults to `["role:write"]`.

A user must have at least one of these permissions in order to use the `brawl merge` command.

### Try Permissions

The `try_permissions` option is a list of permissions required to try a PR.

This defaults to the same as the [**`merge_permissions`**](./repository-configuration.md#merge-permissions) option.

A user must have at least one of these permissions in order to use the `brawl try` command.

### Reviewer Permissions

The `reviewer_permissions` option is a list of permissions required to be a reviewer.

This defaults to the same as the [**`merge_permissions`**](./repository-configuration.md#merge-permissions) option.

A user must have at least one of these permissions in order to be saved as a reviewer when doing a merge.

The reason for this is that on GitHub anyone can review a PR so we can't just use every single person who reviews the PR because they might not be a member of the repository, this is a way to specify which permissions users need in order for us to see them as a valid reviewer.

### Required Status Checks

The `required_status_checks` option is a list of status checks that must be successful before a PR can be merged.

This defaults to `["brawl-done"]`.

If this list is empty, the PR will be merged immediately.
The PR will not be merged until all of the status checks are successful, if any of the status checks fail the PR will immediately fail and report a failure. If all of the status checks are not successful within the [**`timeout_minutes`**](./repository-configuration.md#timeout-minutes) the PR will fail to merge.

### Timeout Minutes

The `timeout_minutes` option is the number of minutes to wait before declaring the merge failed if the required status checks are not met.

This defaults to `60`.

If the required status checks are not met within this time, the PR will fail to merge.

### Max Reviewers

The `max_reviewers` option is the maximum number of reviewers to request for a PR.

This defaults to `10`.


server/src/github/config.rs
use std::fmt::Display;
use std::str::FromStr;

use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[derive(Debug, Deserialize, Serialize, Clone, smart_default::SmartDefault)]
#[serde(default)]
pub struct GitHubBrawlRepoConfig {
    /// Whether Brawl is enabled for this repo
    #[default(true)]
    pub enabled: bool,
    /// Labels to attach to PRs on different states
    pub labels: GitHubBrawlLabelsConfig,
    /// The target branches that this queue matches against.
    pub branches: Vec<String>,
    /// The branch prefix for @brawl try commands
    /// (default: "automation/brawl/try/")
    #[default("automation/brawl/try/")]
    pub try_branch_prefix: String,
    /// The branch prefix for @brawl merge commands
    /// (default: "automation/brawl/merge/")
    #[default("automation/brawl/merge/")]
    pub merge_branch_prefix: String,
    /// The branch prefix for temp branches used when performing merges
    /// (default: "automation/brawl/temp/")
    #[default("automation/brawl/temp/")]
    pub temp_branch_prefix: String,
    /// The permissions required to merge a PR
    /// (default: ["role:write"])
    #[default(vec![Permission::Role(Role::Push)])]
    pub merge_permissions: Vec<Permission>,
    /// The status checks required to merge a PR
    /// (default: ["brawl-done"])
    ///
    /// If brawl will wait for all of these status checks to be successful
    /// before merging. If not provided the PR will be merged instantly.
    #[default(vec![
		"brawl-done".to_string()
	])]
    pub required_status_checks: Vec<String>,
    /// The number of minutes to wait before declaring the merge failed if the
    /// required status checks are not met.
    #[default(60)]
    pub timeout_minutes: i32,
    /// The permissions required to try a commit
    /// (default: <same as merge permissions>)
    #[default(None)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub try_permissions: Option<Vec<Permission>>,
    /// The maximum number of reviewers for a PR
    /// (default: 10)
    #[default(10)]
    pub max_reviewers: i32,
    /// The permissions required to be a reviewer
    /// (default: <same as merge permissions>)
    #[default(None)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reviewer_permissions: Option<Vec<Permission>>,
}

impl GitHubBrawlRepoConfig {
    pub fn missing() -> Self {
        Self {
            enabled: false,
            ..Default::default()
        }
    }

    pub fn try_permissions(&self) -> &[Permission] {
        self.try_permissions.as_ref().unwrap_or(&self.merge_permissions)
    }

    pub fn reviewer_permissions(&self) -> &[Permission] {
        self.reviewer_permissions.as_ref().unwrap_or(&self.merge_permissions)
    }

    pub fn merge_branch(&self, ref_field: &str) -> String {
        format!("{}/{}", self.merge_branch_prefix.trim_end_matches('/'), ref_field)
    }

    pub fn temp_branch(&self) -> String {
        format!("{}/{}", self.temp_branch_prefix.trim_end_matches('/'), uuid::Uuid::new_v4())
    }

    pub fn try_branch(&self, pr_number: u64) -> String {
        format!("{}/{}", self.try_branch_prefix.trim_end_matches('/'), pr_number)
    }
}

#[derive(Debug, Deserialize, Serialize, Clone, smart_default::SmartDefault)]
#[serde(default)]
pub struct GitHubBrawlLabelsConfig {
    /// The label to attach to PRs when they are in the merge queue
    #[serde(skip_serializing_if = "Vec::is_empty", deserialize_with = "string_or_vec")]
    pub on_merge_queued: Vec<String>,
    /// The label to attach to PRs when they are being merged
    #[serde(skip_serializing_if = "Vec::is_empty", deserialize_with = "string_or_vec")]
    pub on_merge_in_progress: Vec<String>,
    /// The label to attach to PRs when they fail to merge
    #[serde(skip_serializing_if = "Vec::is_empty", deserialize_with = "string_or_vec")]
    pub on_merge_failure: Vec<String>,
    /// The label to attach to PRs when they are merged
    #[serde(skip_serializing_if = "Vec::is_empty", deserialize_with = "string_or_vec")]
    pub on_merge_success: Vec<String>,
    /// The label to attach to PRs when they are being tried
    #[serde(skip_serializing_if = "Vec::is_empty", deserialize_with = "string_or_vec")]
    pub on_try_in_progress: Vec<String>,
    /// The label to attach to PRs when they fail to try
    #[serde(skip_serializing_if = "Vec::is_empty", deserialize_with = "string_or_vec")]
    pub on_try_failure: Vec<String>,
    /// The label attached to PRs when auto-try is enabled
    #[serde(skip_serializing_if = "Vec::is_empty", deserialize_with = "string_or_vec")]
    pub auto_try_enabled: Vec<String>,
}

fn string_or_vec<'de, D: Deserializer<'de>>(s: D) -> Result<Vec<String>, D::Error> {
    use serde::de::SeqAccess;

    struct StringOrVecVisitor;

    impl<'de> serde::de::Visitor<'de> for StringOrVecVisitor {
        type Value = Vec<String>;

        #[cfg_attr(coverage_nightly, coverage(off))]
        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
            formatter.write_str("a string or a vector of strings")
        }

        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E> {
            Ok(vec![v.to_string()])
        }

        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
        where
            A: SeqAccess<'de>,
        {
            let mut vec = Vec::new();
            while let Some(value) = seq.next_element()? {
                vec.push(value);
            }
            Ok(vec)
        }
    }

    s.deserialize_any(StringOrVecVisitor)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Permission {
    Role(Role),
    Team(String),
    User(String),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Role {
    Pull,
    Push,
    Admin,
    Maintain,
    Triage,
}

impl From<Role> for octocrab::params::teams::Permission {
    fn from(role: Role) -> Self {
        match role {
            Role::Pull => Self::Pull,
            Role::Push => Self::Push,
            Role::Admin => Self::Admin,
            Role::Maintain => Self::Maintain,
            Role::Triage => Self::Triage,
        }
    }
}

impl FromStr for Role {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "pull" | "read" => Self::Pull,
            "push" | "write" => Self::Push,
            "triage" => Self::Triage,
            "maintain" => Self::Maintain,
            "admin" => Self::Admin,
            _ => return Err(()),
        })
    }
}

impl Display for Role {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Role::Pull => write!(f, "pull"),
            Role::Push => write!(f, "push"),
            Role::Admin => write!(f, "admin"),
            Role::Maintain => write!(f, "maintain"),
            Role::Triage => write!(f, "triage"),
        }
    }
}

impl FromStr for Permission {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (prefix, value) = s.split_once(':').ok_or(())?;

        Ok(match prefix {
            "role" => Self::Role(value.parse().map_err(|_| ())?),
            "team" => Self::Team(value.to_string()),
            "user" => Self::User(value.to_string()),
            _ => return Err(()),
        })
    }
}

impl Display for Permission {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Permission::Role(role) => write!(f, "role:{}", role),
            Permission::Team(team) => write!(f, "team:{}", team),
            Permission::User(user) => write!(f, "user:{}", user),
        }
    }
}

impl<'de> Deserialize<'de> for Permission {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        s.parse().map_err(|_| serde::de::Error::custom("invalid permission"))
    }
}

impl Serialize for Permission {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
mod tests {
    use super::*;

    #[test]
    fn test_from_str() {
        let cases = [
            ("role:pull", Ok(Permission::Role(Role::Pull))),
            ("role:push", Ok(Permission::Role(Role::Push))),
            ("role:read", Ok(Permission::Role(Role::Pull))),
            ("role:write", Ok(Permission::Role(Role::Push))),
            ("role:admin", Ok(Permission::Role(Role::Admin))),
            ("role:maintain", Ok(Permission::Role(Role::Maintain))),
            ("role:triage", Ok(Permission::Role(Role::Triage))),
            ("role:unknown", Err(())),
            ("team:brawl", Ok(Permission::Team("brawl".to_string()))),
            ("user:brawl", Ok(Permission::User("brawl".to_string()))),
            ("team:brawl:brawl", Ok(Permission::Team("brawl:brawl".to_string()))),
            ("user:brawl:brawl", Ok(Permission::User("brawl:brawl".to_string()))),
            ("weird:brawl", Err(())),
            ("invalid", Err(())),
        ];

        for (input, expected) in cases {
            assert_eq!(input.parse::<Permission>(), expected);
        }
    }

    #[test]
    fn test_deserialize() {
        let cases = [
            ("\"role:pull\"", Ok(Permission::Role(Role::Pull))),
            ("\"team:brawl\"", Ok(Permission::Team("brawl".to_string()))),
            ("\"user:brawl\"", Ok(Permission::User("brawl".to_string()))),
            ("\"invalid\"", Err(())),
        ];

        for (input, expected) in cases {
            let result = serde_json::from_str::<Permission>(input);
            match (result, expected) {
                (Ok(result), Ok(expected)) => assert_eq!(result, expected),
                (Err(e), Ok(_)) => panic!("expected ok, got error: {e}"),
                (Ok(_), Err(_)) => panic!("expected error"),
                (Err(_), Err(_)) => {}
            }
        }
    }

    #[test]
    fn test_role_convert() {
        matches!(
            octocrab::params::teams::Permission::from(Role::Pull),
            octocrab::params::teams::Permission::Pull
        );
        matches!(
            octocrab::params::teams::Permission::from(Role::Push),
            octocrab::params::teams::Permission::Push
        );
        matches!(
            octocrab::params::teams::Permission::from(Role::Admin),
            octocrab::params::teams::Permission::Admin
        );
        matches!(
            octocrab::params::teams::Permission::from(Role::Maintain),
            octocrab::params::teams::Permission::Maintain
        );
        matches!(
            octocrab::params::teams::Permission::from(Role::Triage),
            octocrab::params::teams::Permission::Triage
        );
    }

    #[test]
    fn test_missing_disabled() {
        let config = GitHubBrawlRepoConfig::missing();
        assert!(!config.enabled);
    }

    #[test]
    fn test_default_config() {
        let config = GitHubBrawlRepoConfig::default();
        let s = toml::to_string_pretty(&config).unwrap();
        insta::assert_snapshot!(s, @r#"
        enabled = true
        branches = []
        try_branch_prefix = "automation/brawl/try/"
        merge_branch_prefix = "automation/brawl/merge/"
        temp_branch_prefix = "automation/brawl/temp/"
        merge_permissions = ["role:push"]
        required_status_checks = ["brawl-done"]
        timeout_minutes = 60
        max_reviewers = 10

        [labels]
        "#);
    }

    #[test]
    fn test_config_deserialize() {
        let config = r#"
        enabled = true
        branches = [
            "main",
            "staging",
        ]
        try_branch_prefix = "automation/brawl/try/"
        merge_branch_prefix = "automation/brawl/merge/"
        temp_branch_prefix = "automation/brawl/temp/"
        merge_permissions = ["role:push"]
        required_status_checks = ["brawl-done"]
        timeout_minutes = 60
        max_reviewers = 10

        [labels]
        on_merge_queued = "merge-queued"
        on_merge_in_progress = ["merge-in-progress", "waiting-on-brawl"]
        on_merge_failure = "merge-failed"
        on_merge_success = "merge-success"
        on_try_in_progress = ["try-in-progress", "waiting-on-brawl"]
        on_try_failure = "try-failed"
        "#;

        let config: GitHubBrawlRepoConfig = toml::from_str(config).unwrap();
        assert!(config.enabled);
        assert_eq!(config.branches, vec!["main", "staging"]);
        assert_eq!(config.try_branch_prefix, "automation/brawl/try/");
        assert_eq!(config.merge_branch_prefix, "automation/brawl/merge/");
        assert_eq!(config.temp_branch_prefix, "automation/brawl/temp/");
        assert_eq!(config.merge_permissions, vec![Permission::Role(Role::Push)]);
        assert_eq!(config.required_status_checks, vec!["brawl-done"]);
        assert_eq!(config.timeout_minutes, 60);
        assert_eq!(config.max_reviewers, 10);
        assert_eq!(config.labels.on_merge_queued, vec!["merge-queued".to_string()]);
        assert_eq!(
            config.labels.on_merge_in_progress,
            vec!["merge-in-progress".to_string(), "waiting-on-brawl".to_string()]
        );
        assert_eq!(config.labels.on_merge_failure, vec!["merge-failed".to_string()]);
        assert_eq!(config.labels.on_merge_success, vec!["merge-success".to_string()]);
        assert_eq!(
            config.labels.on_try_in_progress,
            vec!["try-in-progress".to_string(), "waiting-on-brawl".to_string()]
        );
        assert_eq!(config.labels.on_try_failure, vec!["try-failed".to_string()]);
    }

    #[test]
    fn test_role_to_string() {
        assert_eq!(Role::Pull.to_string(), "pull");
        assert_eq!(Role::Push.to_string(), "push");
        assert_eq!(Role::Admin.to_string(), "admin");
        assert_eq!(Role::Maintain.to_string(), "maintain");
        assert_eq!(Role::Triage.to_string(), "triage");
    }

    #[test]
    fn test_permission_to_string() {
        assert_eq!(Permission::Role(Role::Pull).to_string(), "role:pull");
        assert_eq!(Permission::Team("brawl".to_string()).to_string(), "team:brawl");
        assert_eq!(Permission::User("brawl".to_string()).to_string(), "user:brawl");
    }

    #[test]
    fn test_try_permissions() {
        let config = GitHubBrawlRepoConfig::default();
        assert_eq!(config.try_permissions(), &config.merge_permissions);
        let config = GitHubBrawlRepoConfig {
            try_permissions: Some(vec![Permission::Role(Role::Push)]),
            ..Default::default()
        };
        assert_eq!(config.try_permissions(), config.try_permissions.as_deref().unwrap());
    }

    #[test]
    fn test_reviewer_permissions() {
        let config = GitHubBrawlRepoConfig::default();
        assert_eq!(config.reviewer_permissions(), &config.merge_permissions);
        let config = GitHubBrawlRepoConfig {
            reviewer_permissions: Some(vec![Permission::Role(Role::Push)]),
            ..Default::default()
        };
        assert_eq!(config.reviewer_permissions(), config.reviewer_permissions.as_deref().unwrap());
    }

    #[test]
    fn test_merge_branch() {
        let config = GitHubBrawlRepoConfig::default();
        assert_eq!(config.merge_branch("main"), "automation/brawl/merge/main");
        let config = GitHubBrawlRepoConfig {
            merge_branch_prefix: "automation/brawl/merge-2a".to_string(),
            ..Default::default()
        };
        assert_eq!(config.merge_branch("main"), "automation/brawl/merge-2a/main");
        let config = GitHubBrawlRepoConfig {
            merge_branch_prefix: "automation/brawl/merge-3a///////////".to_string(),
            ..Default::default()
        };
        assert_eq!(config.merge_branch("main"), "automation/brawl/merge-3a/main");
    }

    #[test]
    fn test_temp_branch() {
        let config = GitHubBrawlRepoConfig::default();
        assert!(config.temp_branch().starts_with("automation/brawl/temp/"));
        let config = GitHubBrawlRepoConfig {
            temp_branch_prefix: "automation/brawl/temp-2a".to_string(),
            ..Default::default()
        };
        assert!(config.temp_branch().starts_with("automation/brawl/temp-2a/"));
        let config = GitHubBrawlRepoConfig {
            temp_branch_prefix: "automation/brawl/temp-3a///////////".to_string(),
            ..Default::default()
        };
        assert!(config.temp_branch().starts_with("automation/brawl/temp-3a/"));
    }

    #[test]
    fn test_try_branch() {
        let config = GitHubBrawlRepoConfig::default();
        assert_eq!(config.try_branch(123), "automation/brawl/try/123");
        let config = GitHubBrawlRepoConfig {
            try_branch_prefix: "automation/brawl/try-2a".to_string(),
            ..Default::default()
        };
        assert_eq!(config.try_branch(123), "automation/brawl/try-2a/123");
        let config = GitHubBrawlRepoConfig {
            try_branch_prefix: "automation/brawl/try-3a///////////".to_string(),
            ..Default::default()
        };
        assert_eq!(config.try_branch(123), "automation/brawl/try-3a/123");
    }
}
